File: macos/AirNinjaCore/Sources/AirNinjaCore/Base32.swift
import Foundation

public enum Base32 {
    private static let alphabet = Array("abcdefghijklmnopqrstuvwxyz234567")
    private static let bitsPerSymbol = 5
    private static let byteBits = 8
    private static let symbolMask = 0x1F

    public static func encodeLowerNoPadding(_ data: Data) -> String {
        var result = ""
        var buffer = 0
        var bitsLeft = 0
        for byte in data {
            buffer = (buffer << byteBits) | Int(byte)
            bitsLeft += byteBits
            while bitsLeft >= bitsPerSymbol {
                let index = (buffer >> (bitsLeft - bitsPerSymbol)) & symbolMask
                result.append(alphabet[index])
                bitsLeft -= bitsPerSymbol
            }
        }
        if bitsLeft > 0 {
            let index = (buffer << (bitsPerSymbol - bitsLeft)) & symbolMask
            result.append(alphabet[index])
        }
        return result
    }
}

File: protocol-core/src/main/kotlin/com/airninja/protocol/crypto/Base32.kt
package com.airninja.protocol.crypto

object Base32 {
    private const val ALPHABET = "abcdefghijklmnopqrstuvwxyz234567"
    private const val BITS_PER_SYMBOL = 5
    private const val BYTE_BITS = 8
    private const val SYMBOL_MASK = 0x1F

    fun encodeLowerNoPadding(data: ByteArray): String {
        val builder = StringBuilder()
        var buffer = 0
        var bitsLeft = 0
        for (byte in data) {
            buffer = (buffer shl BYTE_BITS) or (byte.toInt() and 0xFF)
            bitsLeft += BYTE_BITS
            while (bitsLeft >= BITS_PER_SYMBOL) {
                val index = (buffer shr (bitsLeft - BITS_PER_SYMBOL)) and SYMBOL_MASK
                builder.append(ALPHABET[index])
                bitsLeft -= BITS_PER_SYMBOL
            }
        }
        if (bitsLeft > 0) {
            val index = (buffer shl (BITS_PER_SYMBOL - bitsLeft)) and SYMBOL_MASK
            builder.append(ALPHABET[index])
        }
        return builder.toString()
    }
}
